blog/_posts/2023-12-08-migration-to-scala-3.md

```diff
@@ -0,0 +1,146 @@
+- What external libraries do we use? Do they have a Scala3-compatible release?
+
+Other questions you might need to consider:
+
+- How many services do you need to migrate?
+- What Scala version(s) does each service use?
+
+Thanks to the [interoperability](https://docs.scala-lang.org/scala3/guides/migration/compatibility-source.html) between Scala 3 and Scala 2.13, it might be more accessible or crucial to upgrade to Scala 2.13 first. You have to keep this in mind.
+
+*By this point, we knew that all our services were on 2.13; we knew how many internal libraries needed to be used from both Scala versions and what external libraries had no Scala 3 support: what is the status, what is missing, and what can we do to deal with it. For example, a couple of libraries got a Scala 3 release by the time we were done with the investigation, and one `circe` ”extras” library was replaced with two lines of boilerplate.*
+
+## Step 2. Draw the rest of the owl
+
+After all the meetings and/or writings comes the time to do the actual work. When you have concrete questions or action items, remember that [Scala 3 migration guides](https://docs.scala-lang.org/scala3/guides/migration/compatibility-intro.html) are your friends. For example, it covers how to port compiler options and sbt projects. Great reference resource.
+
+We don’t have to go into these details right now. Instead, let’s talk about the bigger picture and a few other things to remember.
+
+### Dealing with internal libraries
+
+There are multiple ways of sharing a library between Scala 2 and Scala 3 services.
+
+**The good news** is that you can use Scala 2.13 libraries from Scala 3 app and Scala 3 libraries from Scala 2.13 app.
+
+**The bad news** is that you can’t do it directly if you have shared transitive dependencies.
+
+**The good news** is that it’s pretty easy to deal with.
+
+Imagine you have an internal library (e.g., auth-handling, specific DB-connector, peculiar API) used by your existing Scala 2 services that will also be used by Scala 3 services. If your library on Scala 2 uses the `ducks` library and your Scala 3 application uses `ducks` as well, you get two conflicting versions `ducks_2.13` and `ducks_3`.
+
+To solve this, you can cross-publish your library to both versions by adding one line to your build:
+
+```scala
++ crossScalaVersions := List("2.13.12", "3.3.1")
+```
+
+To build against all versions listed in `crossScalaVersions`, prefix the action to run with `+`. For example:
+
+```
+sbt> + compile
+```
+
+*At least, that’s what we did — everything else just worked. Your mileage may vary.*
+
+> For more information, see the guides on [ClassPath compatibility](https://docs.scala-lang.org/scala3/guides/migration/compatibility-classpath.html) and [Cross-building (sbt)](https://www.scala-sbt.org/1.x/docs/Cross-Build.html) or [Cross Builds (mill)](https://mill-build.com/mill/0.9.12/Cross_Builds.html)
```

Comment:
Mill link may be a bit out of context here as it's the only time you mention Mill in the article - everywhere else you seem to be talking from sbt context.

Reply:
I removed another mention of sbt. Now, both are only mentioned here with an explicit and specific sbt example.